osu.Game/Audio/PreviewTrackManager.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using osu.Framework.Allocation;
+using osu.Framework.Audio;
+using osu.Framework.Audio.Track;
+using osu.Framework.Configuration;
+using osu.Framework.Graphics;
+using osu.Framework.IO.Stores;
+using osu.Game.Beatmaps;
+
+namespace osu.Game.Audio
+{
+    /// <summary>
+    /// A central store for the retrieval of <see cref="PreviewTrack"/>s.
+    /// </summary>
+    public class PreviewTrackManager : Component
+    {
+        private readonly BindableDouble muteBindable = new BindableDouble();
+
+        private AudioManager audio;
+        private TrackManager trackManager;
+
+        private TrackManagerPreviewTrack current;
+
+        [BackgroundDependencyLoader]
+        private void load(AudioManager audio, FrameworkConfigManager config)
+        {
+            trackManager = new TrackManager(new OnlineStore());
+
+            this.audio = audio;
+            audio.AddItem(trackManager);
+
+            config.BindWith(FrameworkSetting.VolumeMusic, trackManager.Volume);
+        }
+
+        /// <summary>
+        /// Retrieves a <see cref="PreviewTrack"/> for a <see cref="BeatmapSetInfo"/>.
+        /// </summary>
+        /// <param name="beatmapSetInfo">The <see cref="BeatmapSetInfo"/> to retrieve the preview track for.</param>
+        /// <returns>The playable <see cref="PreviewTrack"/>.</returns>
+        public PreviewTrack Get(BeatmapSetInfo beatmapSetInfo)
+        {
+            var track = CreatePreviewTrack(beatmapSetInfo, trackManager);
+
+            track.Started += () =>
+            {
+                current?.Stop();
+                current = track;
+                audio.Track.AddAdjustment(AdjustableProperty.Volume, muteBindable);
+            };
+
+            track.Stopped += () =>
+            {
+                current = null;
+                audio.Track.RemoveAdjustment(AdjustableProperty.Volume, muteBindable);
+            };
+
+            return track;
+        }
+
+        /// <summary>
+        /// Stops the currently playing <see cref="PreviewTrack"/>.
+        /// </summary>
+        /// <remarks>
+        /// Only the immediate owner (an object that implements <see cref="IPreviewTrackOwner"/>) of the playing <see cref="PreviewTrack"/>
+        /// can globally stop the currently playing <see cref="PreviewTrack"/>. The object holding a reference to the <see cref="PreviewTrack"/>
+        /// can always stop the <see cref="PreviewTrack"/> themselves through <see cref="PreviewTrack.Stop()"/>.
+        /// </remarks>
+        /// <param name="source">The <see cref="IPreviewTrackOwner"/> which may be the owner of the <see cref="PreviewTrack"/>.</param>
+        public void Stop(IPreviewTrackOwner source)
+        {
+            if (current?.Owner != source)
+                return;
+
+            current?.Stop();
+            current = null;
```

Comment:
weird that current is set twice: once in the Stop call's callback and once here. also i'd prefer an early `if (current == null || current.Owner != source)` return rather than the second `current?` null checks.

Reply:
Yeah I didn't quite want to make a `private stop()` method that is called from `event Stopped` to avoid it looking like an infinite loop.